oxenss/snode/service_node.cpp

```diff
@@ -43,29 +44,426 @@ using MISSING_PUBKEY_THRESHOLD = std::ratio<3, 100>;
 /// TODO: there should be config.h to store constants like these
 constexpr auto OXEND_PING_INTERVAL = 30s;
 
+// How often to trigger 'check_new_members' which checks for 'data ready' handshakes from
+// swarm members and propagate a DB dump if necessary.
 constexpr auto NEW_SWARM_MEMBER_INTERVAL = 10s;
 
+struct SerialiseRetryableRequestsResult {
+    SerialiseBTResult bt;
+    std::vector<RequestRetry> retryable_requests;
+};
+
+SerialiseDataReadyRequestResult serialise_data_ready_request(
+        Serialise serialise, std::string_view read_data, const DataReadyRequest& write_data) {
+    SerialiseDataReadyRequestResult result = {};
+    uint32_t version = 0;
+    constexpr std::string_view VERSION_KEY = "@";
+    constexpr std::string_view STATUS_KEY = "s";
+    constexpr std::string_view NEED_DB_DUMP_KEY = "t";
+    static_assert(VERSION_KEY < STATUS_KEY);
+    static_assert(STATUS_KEY < NEED_DB_DUMP_KEY);
+
+    if (serialise == Serialise::Write) {
+        oxenc::bt_dict_producer d;
+        d.append(VERSION_KEY, version);
+        d.append(NEED_DB_DUMP_KEY, write_data.needs_db_dump);
+        result.bt.write_payload = d.view();
+        result.bt.success = result.bt.error.empty();
+    } else {
+        if (read_data.size()) {
+            oxenc::bt_dict_consumer d{read_data};
+            try {
+                version = d.require<uint8_t>(VERSION_KEY);
+            } catch (const std::exception& e) {
+                result.bt.error =
+                        "Failed to parse sn data ready request version: {}"_format(e.what());
+            }
+
+            if (result.bt.error.empty()) {
+                try {
+                    result.request.needs_db_dump = d.require<bool>(NEED_DB_DUMP_KEY);
+                } catch (const std::exception& e) {
+                    result.bt.error =
+                            "Failed to parse sn data ready db dump flag: {}"_format(e.what());
+                }
+            }
+        } else {
+            result.bt.error = "Failed to parse data ready payload: no bytes given";
+        }
+
+        result.bt.success = result.bt.error.empty();
+    }
+    return result;
+}
+
+static SerialiseRetryableRequestsResult serialize_retryable_requests(
+        Serialise serialise, std::string_view read_data, std::span<RequestRetry> write_data) {
+    SerialiseRetryableRequestsResult result = {};
+    uint32_t version = 0;
+
+    constexpr std::string_view VERSION_KEY = "@";
+    constexpr std::string_view RETRYABLE_REQUESTS_KEY = "r";
+    assert(VERSION_KEY < RETRYABLE_REQUESTS_KEY);
+
+    // Retryable request keys
+    constexpr std::string_view COMMAND_KEY = "c";
+    constexpr std::string_view REQ_PAYLOAD_KEY = "r";
+    constexpr std::string_view CREATE_TIME_KEY = "t";
+    constexpr std::string_view NODES_KEY = "u";
+    assert(COMMAND_KEY < REQ_PAYLOAD_KEY);
+    assert(REQ_PAYLOAD_KEY < CREATE_TIME_KEY);
+    assert(CREATE_TIME_KEY < NODES_KEY);
```

Comment:
Checks like this that can be `static_assert`s should be.  Save `assert` for invariants only testable at runtime.